Source/WebCore/svg/SVGPathElement.cpp

```diff
@@ -37,15 +37,59 @@ namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(SVGPathElement);
 
-static constexpr uint64_t maxPathSegListCacheSize = 150 * 1024; // 150 Kb.
-static uint64_t pathSegListCacheSize = 0;
+class PathSegListCache {
+public:
+    static PathSegListCache& singleton();
 
-static HashMap<AtomString, SVGPathByteStream::Data>& pathSegListCache()
+    const SVGPathByteStream::Data* get(const AtomString& attributeValue) const;
+    void add(const AtomString& attributeValue, const SVGPathByteStream::Data&);
+    void clear();
+
+private:
+    friend class NeverDestroyed<PathSegListCache, MainThreadAccessTraits>;
+    PathSegListCache() = default;
+
+    HashMap<AtomString, SVGPathByteStream::Data> m_cache;
+    uint64_t m_sizeInBytes { 0 };
+    static constexpr uint64_t maxSizeInBytes = 150 * 1024; // 150 Kb.
+};
+
+PathSegListCache& PathSegListCache::singleton()
 {
-    static MainThreadNeverDestroyed<HashMap<AtomString, SVGPathByteStream::Data>> cache;
+    static MainThreadNeverDestroyed<PathSegListCache> cache;
     return cache;
 }
 
+const SVGPathByteStream::Data* PathSegListCache::get(const AtomString& attributeValue) const
+{
+    auto iterator = m_cache.find(attributeValue);
+    return iterator != m_cache.end() ? &iterator->value : nullptr;
+}
+
+void PathSegListCache::add(const AtomString& attributeValue, const SVGPathByteStream::Data& data)
+{
+    size_t newDataSize = data.size();
+    if (UNLIKELY(newDataSize > (maxSizeInBytes / 2)))
+        return;
+
+    m_sizeInBytes += newDataSize;
```

Comment:
Should we consider adding small size paths only? Some SVGs like maps use very long path elements. I think this cache will be filled with small number of large paths in this case.

Reply:
Yes, it is worth trying out but needs careful A/B testing on Speedometer 3.